src/tree/mod.rs

```diff
@@ -591,6 +598,19 @@
         Ok(())
     }
 
+    /// Add a node for the provided key package the tree on the right side.
+    /// Note, that this function will not fill blank leaves.
+    fn add_node(&mut self, key_package: &KeyPackage) -> (NodeIndex, Credential) {
+        if !self.nodes.is_empty() {
+            self.nodes.push(Node::new_blank_parent_node());
+        }
+        self.nodes.push(Node::new_leaf(Some((key_package).clone())));
```

Comment:
This `clone` shouldn't be necessary. You have to change `add_nodes` not to take a reference. But then I don't see a need for cloning.

Reply:
Well, the KeyPackage references get down here all the way from the proposals. I could try to refactor how proposals are handled, but I gave that a go and found that it's not straight-forward. Would you prefer if I have the caller do the cloning?

Reply:
I guess the only way to get rid of the `clone()` would be to consume the `KeyPackage`. Not sure if that works for the caller though.

Reply:
I don't see a reason why it shouldn't work. `add_nodes` gets called with a list of key package references. It should be a list of key packages instead, i.e. the `add_proposals.iter()` there should actually take out the elements instead of referencing. This would also avoid any potential bug of re-using key packages because of wrong indices.
But this would actually require to refactor the way the proposals are applied... so let's leave it like this for now as we need the key package in two places right now.

Reply:
That's what I meant. It probably makes sense to do, but this PR already has some serious scope creep 😄